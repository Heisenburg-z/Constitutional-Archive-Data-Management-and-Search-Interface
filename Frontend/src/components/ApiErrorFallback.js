import React from 'react';
import { AlertTriangle, RefreshCw } from 'lucide-react';

/**
 * Component to display when API calls fail
 * 
 * @param {Object} props
 * @param {Function} props.retryAction - Function to call when retry button is clicked
 * @param {string} props.message - Custom error message to display
 * @param {boolean} props.isHomePageError - Whether this is displayed on the homepage (affects styling)
 */
const ApiErrorFallback = ({ retryAction, message, isHomePageError = false }) => {
  return (
    <div className={`${isHomePageError ? 'w-full max-w-2xl mx-auto' : ''} bg-red-50 border border-red-200 rounded-lg p-4 shadow-sm`}>
      <div className="flex items-start">
        <AlertTriangle className="h-5 w-5 text-red-600 mr-3 mt-0.5 flex-shrink-0" />
        <div className="flex-1">
          <h3 className="text-sm font-medium text-red-800">API Connection Error</h3>
          <div className="mt-1 text-sm text-red-700">
            <p>{message || "We're having trouble connecting to our API service."}</p>
            <p className="mt-1">Please check your network connection and API configuration.</p>
          </div>
          <div className="mt-3">
            <button
              type="button"
              onClick={retryAction}
              className="inline-flex items-center px-3 py-2 border border-transparent text-sm leading-4 font-medium rounded-md text-white bg-red-600 hover:bg-red-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-red-500"
            >
              <RefreshCw className="h-4 w-4 mr-1" />
              Retry Connection
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ApiErrorFallback;