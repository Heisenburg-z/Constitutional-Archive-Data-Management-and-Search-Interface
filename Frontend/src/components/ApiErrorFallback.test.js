import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ApiErrorFallback from './ApiErrorFallback';

describe('ApiErrorFallback Component', () => {
  test('renders without crashing', () => {
    render(<ApiErrorFallback />);
    expect(screen.getByText('Unable to connect to the server')).toBeInTheDocument();
  });

  test('displays default error message when no message prop is provided', () => {
    render(<ApiErrorFallback />);
    expect(screen.getByText('Unable to connect to the server')).toBeInTheDocument();
  });

  test('displays custom error message when message prop is provided', () => {
    const customMessage = 'Custom error message';
    render(<ApiErrorFallback message={customMessage} />);
    expect(screen.getByText(customMessage)).toBeInTheDocument();
  });

  test('displays the standard description text', () => {
    render(<ApiErrorFallback />);
    expect(
      screen.getByText('We\'re experiencing some technical difficulties. Please try again later.')
    ).toBeInTheDocument();
  });

  test('does not render retry button when retryAction prop is not provided', () => {
    render(<ApiErrorFallback />);
    const retryButton = screen.queryByText('Retry Connection');
    expect(retryButton).not.toBeInTheDocument();
  });

  test('renders retry button when retryAction prop is provided', () => {
    const retryAction = jest.fn();
    render(<ApiErrorFallback retryAction={retryAction} />);
    const retryButton = screen.getByText('Retry Connection');
    expect(retryButton).toBeInTheDocument();
  });

  test('calls retryAction when retry button is clicked', () => {
    const retryAction = jest.fn();
    render(<ApiErrorFallback retryAction={retryAction} />);
    
    const retryButton = screen.getByText('Retry Connection');
    fireEvent.click(retryButton);
    
    expect(retryAction).toHaveBeenCalledTimes(1);
  });

  test('renders the FileText icon', () => {
    render(<ApiErrorFallback />);
    
    // Test that the container for the icon exists
    const iconContainer = screen.getByText('Unable to connect to the server')
      .parentElement
      .querySelector('.bg-red-100');
    
    expect(iconContainer).toBeInTheDocument();
    expect(iconContainer).toHaveClass('inline-flex');
    expect(iconContainer).toHaveClass('items-center');
    expect(iconContainer).toHaveClass('justify-center');
  });

  test('applies correct styling classes', () => {
    render(<ApiErrorFallback />);
    
    // Main container
    const container = screen.getByText('Unable to connect to the server').closest('div');
    expect(container).toHaveClass('bg-red-50');
    expect(container).toHaveClass('border');
    expect(container).toHaveClass('border-red-200');
    expect(container).toHaveClass('rounded-lg');
    
    // Heading
    const heading = screen.getByText('Unable to connect to the server');
    expect(heading).toHaveClass('text-lg');
    expect(heading).toHaveClass('font-medium');
    expect(heading).toHaveClass('text-gray-900');
    
    // Description
    const description = screen.getByText('We\'re experiencing some technical difficulties. Please try again later.');
    expect(description).toHaveClass('text-gray-600');
  });

  test('renders retry button with RefreshCw icon when retryAction is provided', () => {
    const retryAction = jest.fn();
    render(<ApiErrorFallback retryAction={retryAction} />);
    
    const retryButton = screen.getByText('Retry Connection').closest('button');
    
    // Check button styling
    expect(retryButton).toHaveClass('bg-blue-600');
    expect(retryButton).toHaveClass('hover:bg-blue-700');
    expect(retryButton).toHaveClass('text-white');
    expect(retryButton).toHaveClass('inline-flex');
    expect(retryButton).toHaveClass('items-center');
    
    // Verify button has the right attributes
    expect(retryButton).toHaveAttribute('class', expect.stringContaining('focus:ring-2'));
    expect(retryButton).toHaveAttribute('class', expect.stringContaining('focus:outline-none'));
  });

  test('snapshot test', () => {
    const retryAction = jest.fn();
    const { container } = render(<ApiErrorFallback retryAction={retryAction} message="Test error message" />);
    expect(container).toMatchSnapshot();
  });
});