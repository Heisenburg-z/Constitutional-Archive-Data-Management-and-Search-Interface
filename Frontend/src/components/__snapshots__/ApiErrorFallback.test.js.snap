// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiErrorFallback Component snapshot test 1`] = `
<div>
  <div
    class="bg-red-50 border border-red-200 rounded-lg p-6 mb-8 text-center"
  >
    <div
      class="bg-red-100 inline-flex items-center justify-center w-12 h-12 rounded-full mb-4"
    >
      <svg
        class="lucide lucide-file-text h-6 w-6 text-red-500"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V7Z"
        />
        <path
          d="M14 2v4a2 2 0 0 0 2 2h4"
        />
        <path
          d="M10 9H8"
        />
        <path
          d="M16 13H8"
        />
        <path
          d="M16 17H8"
        />
      </svg>
    </div>
    <h3
      class="text-lg font-medium text-gray-900 mb-2"
    >
      Test error message
    </h3>
    <p
      class="text-gray-600 mb-4"
    >
      We're experiencing some technical difficulties. Please try again later.
    </p>
    <button
      class="inline-flex items-center px-4 py-2 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
    >
      <svg
        class="lucide lucide-refresh-cw h-4 w-4 mr-2"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
        />
        <path
          d="M21 3v5h-5"
        />
        <path
          d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
        />
        <path
          d="M8 16H3v5"
        />
      </svg>
      Retry Connection
    </button>
  </div>
</div>
`;
